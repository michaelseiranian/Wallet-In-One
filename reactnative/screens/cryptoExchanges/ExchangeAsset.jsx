import {
    Image,
    Pressable,
    StyleSheet,
    Text,
    TouchableOpacity,
    TouchableWithoutFeedback,
    View,
  } from "react-native";
import React, {useEffect, useState, useCallback} from "react";
import getCryptoIcon from "../cryptocurrency/icons/icon";
import { useTheme } from 'reactnative/src/theme/ThemeProvider';
import { api_url } from "../../authentication";
import * as SecureStore from 'expo-secure-store';

export default function ExchangeAsset(props) {
  // Dark mode variables
  const {dark, colors, setScheme} = useTheme();

  const styles = StyleSheet.create({
    exchangeAsset: {
      padding: 10,
      marginVertical: 5,
      borderRadius: 10,
      flexDirection: "row",
    },
    exchangeAssetTitle: {
      fontWeight: "700",
      flex: 1,
    },
    exchangeAssetImage: {
      width: 30,
      height: 30,
    },
  });
    
  
    // Render script
    return (
        // Navigate to ExchangeTransactions on press
      <TouchableWithoutFeedback
        onPress={() =>
          props.navigation.navigate("ExchangeTransactions",
            { item: props.item, removeExchange: props.removeExchange})
        }
        testID="exchangeAsset"
      >
        <View style={[styles.exchangeAsset, {backgroundColor: colors.primary}]}>
          <View
            style={{
              alignItems: "center",
              justifyContent: "center",
              paddingRight: 10,
            }}
          >
              {/*Load crypto exchange logo*/}
            <Image
              style={styles.exchangeAssetImage}
              source={getCryptoIcon(props.item.crypto_exchange_name)}
            />
          </View>
  
          <View
            style={{
              flex: 1,
              flexDirection: "row",
              justifyContent: "space-between",
            }}
          >
              {/*Write the name on the asset card*/}
            <View style={{}}>
              <Text style={{ fontSize: 25, fontWeight: "700", color: colors.text }}>
                {props.item.crypto_exchange_name}
              </Text>
                {/*Show loading on Total balance if empty, else show the actual balance*/}
              {props.balances.length == 0 ? (
                <Text style={[styles.exchangeAssetTitle, {color: colors.text}]}>Total Balance: £Loading...</Text>
              ) : (
                <Text style={[styles.exchangeAssetTitle, {color: colors.text}]}>
                  Total Balance: £{props.balances.find(balance => balance.id === props.item.id)?.y.toFixed(2) || '0.00'}
                </Text>
              )}
            </View>
  
          </View>
        </View>
      </TouchableWithoutFeedback>
    );
  }