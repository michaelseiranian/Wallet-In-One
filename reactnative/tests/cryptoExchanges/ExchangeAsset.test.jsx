import { render, fireEvent } from '@testing-library/react-native';
import ExchangeAsset from '../../screens/cryptoExchanges/ExchangeAsset';
import renderer from 'react-test-renderer';

describe('ExchangeAsset component', () => {
    it('renders correctly', () => {
      const props = {
        item: { id: 1, crypto_exchange_name: 'Test Exchange' },
        balances: [],
        navigation: {
            navigate: jest.fn()
        }
      };  
      const tree = renderer.create(<ExchangeAsset {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
});

describe('<ExchangeAsset />', () => {
  test('renders correctly with minimum props', () => {
    const props = {
      item: { id: 1, crypto_exchange_name: 'Test Exchange' },
      balances: [],
      navigation: {
        navigate: jest.fn()
      }
    };
    const { getByText, getByTestId } = render(<ExchangeAsset {...props} />);
    
    expect(getByTestId('exchangeAsset')).toBeDefined();
    expect(getByText('Test Exchange')).toBeDefined();
    expect(getByText('Total Balance: £Loading...')).toBeDefined();
  });
});

describe('<ExchangeAsset />', () => {
    test('displays total balance when provided with balances', () => {
      const props = {
        item: { id: 1, crypto_exchange_name: 'Test Exchange' },
        balances: [{ id: 1, x: '2023-03-19T09:00:00.000Z', y: 123.45 }],
        navigation: {
          navigate: jest.fn()
        }
      };
      const { getByText } = render(<ExchangeAsset {...props} />);
      
      expect(getByText('Total Balance: £123.45')).toBeDefined();
    });
  });

describe('<ExchangeAsset />', () => {
  test('navigates to "ExchangeTransactions" screen on press', () => {
    const props = {
      item: { id: 1, crypto_exchange_name: 'Test Exchange' },
      balances: [],
      removeExchange: jest.fn(),
      navigation: {
        navigate: jest.fn()
      }
    };
    const { getByTestId } = render(<ExchangeAsset {...props} />);
    const exchangeAsset = getByTestId('exchangeAsset');
    fireEvent.press(exchangeAsset);
    
    expect(props.navigation.navigate).toHaveBeenCalledWith('ExchangeTransactions', {
      item: props.item,
      removeExchange: props.removeExchange
    });
  });
});