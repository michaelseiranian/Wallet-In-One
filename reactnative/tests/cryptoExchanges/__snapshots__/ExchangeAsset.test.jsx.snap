// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExchangeAsset component renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "borderRadius": 10,
        "flexDirection": "row",
        "marginVertical": 5,
        "padding": 10,
      },
      {
        "backgroundColor": "purple",
      },
    ]
  }
  testID="exchangeAsset"
>
  <View
    style={
      {
        "alignItems": "center",
        "justifyContent": "center",
        "paddingRight": 10,
      }
    }
  >
    <Image
      style={
        {
          "height": 30,
          "width": 30,
        }
      }
    />
  </View>
  <View
    style={
      {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      style={{}}
    >
      <Text
        style={
          {
            "color": "black",
            "fontSize": 25,
            "fontWeight": "700",
          }
        }
      >
        Test Exchange
      </Text>
      <Text
        style={
          [
            {
              "flex": 1,
              "fontWeight": "700",
            },
            {
              "color": "black",
            },
          ]
        }
      >
        Total Balance: Â£Loading...
      </Text>
    </View>
  </View>
</View>
`;
